cnd/src/btsieve/ethereum/mod.rs

```diff
@@ -24,16 +20,134 @@ use std::collections::HashSet;
 type Hash = H256;
 type Block = crate::ethereum::Block<Transaction>;
 
-pub async fn matching_transaction<C>(
+pub const TRANSACTION_STATUS_OK: u32 = 1;
+
+pub async fn matching_create_contract_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    bytecode: Bytes,
+) -> anyhow::Result<Transaction>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction(blockchain_connector.clone(), start_of_swap, |transaction| {
+        Box::new({
+            let blockchain_connector = blockchain_connector.clone();
+            let bytecode = bytecode.clone();
+            async move {
+                // transaction.to address is None if, and only if, the transaction
+                // creates a contract.
+                if transaction.to.is_none() && transaction.input == bytecode {
+                    // We can get a transaction that matches even for failed attempts
+                    // at contract creation e.g., low gas.
+                    let receipt = fetch_receipt(blockchain_connector, transaction.hash).await?;
+                    if receipt.status == Some(TRANSACTION_STATUS_OK.into()) {
+                        return Ok(true);
+                    }
+                }
+                Ok(false)
+            }
+            .boxed()
+        })
+    })
+    .await
+}
+
+pub async fn matching_event_transaction<C>(
+    blockchain_connector: C,
+    start_of_swap: NaiveDateTime,
+    event: Event,
+    action: &'static str,
+) -> anyhow::Result<(Transaction, TransactionReceipt)>
+where
+    C: LatestBlock<Block = Option<Block>>
+        + BlockByHash<Block = Option<Block>, BlockHash = Hash>
+        + ReceiptByHash<Receipt = Option<TransactionReceipt>, TransactionHash = Hash>
+        + Clone,
+{
+    matching_transaction_and_receipt(
+        blockchain_connector.clone(),
+        start_of_swap,
+        event.topics.clone(),
+        action,
+        |transaction| {
+            Box::new({
+                let blockchain_connector = blockchain_connector.clone();
+                let event = event.clone();
+                async move {
+                    let receipt = fetch_receipt(blockchain_connector, transaction.hash).await?;
```

Comment:
Moving this one layer down would make the matcher non-async (and hence a lot more concise).